# CatWalk

![example workflow](https://github.com/dvolk/catwalk/actions/workflows/test_cw.yml/badge.svg)

cw_webui font page             |  cw_webui sample page
:-------------------------:|:-------------------------:
![cw_webui screenshot](https://i.imgur.com/nCdxbXB.png)| ![cw_webui screenshot sample](https://i.imgur.com/hWqTxek.png)

## Description

CatWalk is a memory-efficient server application to compute pairwise single nucleotide polymorphism (SNP) differences between microbial genomes.  It does not require any prior information (such as phylogenetics, lineage, or multi-locus sequence typing) to do so, and can be applied to any fixed length sequence set, such as those generated by read mapping to a reference sequence and consensus sequence calling.

CatWalk offers
* fast, one-time sequence loading from a fasta file
* the ability to add new samples
* ability to query for similar sequences
* access via command line or REST-API interfaces, allowing incorporation into other components.
* a web server interface, so individuals can look up results using the web server  included in the package

## Performance
Catwalk queries are fast:
* To compare one SARS-CoV-2 sample against another takes about 0.4 microseconds.
* To compare one *M. tuberculosis* sample against another takes about 0.7 microseconds.
* So to find close neighbours of a sequence from a million sequences takes under a second.

The speed of comparison depends on the distance of the test samples from the reference, and the number of unknown bases in the sequence.  Simulated data illustrating this is [available](benchmark/sim)

## Context
There is a public health need to rapidly identify sequences which may form part of transmission chains, or which may represent importations across national boundaries amidst large numbers of SARS-CoV-2 and other microbial sequences.

Reference based compression is a sequence analysis technique which allows both compact storage of sequence data and comparisons between sequences.

This software is a fast, memory efficient implementation of reference based compression, comparison and in-RAM storage.

## Installing the software

* The system has only been tested on linux
* nim >= 1.4.8

### Installing nim (on Linux)

The following assumes you're putting nim in the home directory, but you can put it anywhere and then edit the paths below

1. Nim requires gcc:
```
sudo apt install gcc
```

2. Download and extract nim:

```
cd
wget https://nim-lang.org/download/nim-1.4.8-linux_x64.tar.xz
tar xf nim-1.4.8-linux_x64.tar.xz
```
note: also tested successfully with 1.6.0

3. to be able to run nim from any directory, add it to the `PATH` variable (replace yourusername with your username):
```
echo 'PATH=$PATH:/home/yourusername/nim-1.4.8/bin' >> .bashrc
```

4. When you run `nimble install` (the nim package manager), it puts binaries into `~/.nimble/bin`, so add that too (replace yourusername with your username):
```
echo 'PATH=$PATH:/home/yourusername/.nimble/bin' >> .bashrc
```

5. Source the bashrc file:
```
source ~/.bashrc
```
6. Confirm it's working:
```
$ nim -v
Nim Compiler Version 1.4.8 [Linux: amd64]
```

### Building (Compiling) Catwalk

in the catwalk directory, run one of these:
**option 1**: the catwalk server will keep a record of sequences loaded into it on disc, and automatically reload on startup.  See *Cache* section below.

    nimble -y build -d:release -d:danger

or
**option 2**: the catwalk server will not keep a record of files loaded into it.  On restart, the client will need to reload records

    nimble -y build -d:release -d:danger -d:no_serialisation


Both options creates four binaries: `cw_server`, `cw_client`, `cw_webui`, `refcompress`.

Catwalk is a server application.  The Catwalk server `cw_server` stores the sequences and performs distance computations.

The `cw_client` and `cw_webui`, applications target the server using a command line and Web User interface ,respectively.

`refcompress` is a helper application, see below.

## Using Catwalk

### Starting the server (cw_server)

Example:

    ./cw_server --instance-name=test \
                --reference-filepath=reference/nc_045512.fasta \
                --mask-filepath=reference/covid-exclude.txt \

### Unit tests

Using a python virtual environment, run tests through python client [TODO]

    pipenv install
    ./run_tests.sh

### Batch adding from multifasta file (recommended)

If you have a multifasta file (where sequences don't contain newlines, i.e. all sequences are present on one line), you can add the samples with the fast command line client:

    ./cw_client add_samples_from_mfsl -f /home/dv/cog_all.fasta

An alternative is to write a script to load samples via the REST API /add_sample endpoint, see below.
'pycw_client.py' contains a python interface to cw_server.

### Batch adding individual samples

use something like `find`. e.g.:

    find path/to/your/fasta/files -type f -exec ./cw_client add_sample -f {} \;

### Starting the web UI

    ./cw_webui

Open browser at `http://localhost:5001`

### Cache

When you add samples to catwalk, a JSON string of the reference compressed sequence is saved to the `instance-name` directory. When you restart catwalk with that instance name, it will load these sequences automatically. This is much faster than re-adding them every time, and the files are smaller than the original fasta files.

We also provide the `refcompress` binary, which is a program to produce these reference compressed sequences. This enables the reference compression to be done in parallel (e.g. with GNU parallel) or integrated into pipelines.

## Utilities

### pyclient/pycw_client.py

A python interface to Catwalk, which is linux-specific.  Supports starting and stopping catwalk instances,
and interactions with all endpoints exposed by the server.

### utils/make_mfsl.py

Convert a directory of fasta files into the multifasta singleline format.

### utils/verify.py

Given a reference, a mask file and two samples, check the difference between
the samples in different ways.

### utils/compare_neighbours.py

Get all neighbours from different relateness services and compare them. Useful for
comparing identical datasets

### utils/clusters.py

Use networkx to create a cluster graph

#### Example: making a cluster graph png

You will need a running catwalk server and graphviz installed

Add your samples to catwalk and run:

```
python3 compare_neighbours.py cwn 12 > example.txt
python3 clusters.py example.txt 12
sfdp example.dot -Tpng -o example.png
```

This creates a cluster graph for SNP distance 12, see [an output example](https://gitea.mmmoxford.uk/dvolk/catwalk/raw/branch/master/doc/cluster-example.png).

## HTTP API

Catwalk provides the following HTTP API endpoints (examples are given with the Python Requests library):

### /info

Returns server information such which reference is used, memory used, version, etc.

    >>> requests.get("http://localhost:5000/info").json()

### /list_samples

Returns a JSON array of sample names loaded into the server.

    >>> requests.get("http://localhost:5000/list_samples").json()

### /add_sample

Add a sample to catwalk

    >>> requests.post("http://localhost:5000/add_sample", json={ "name": sample_name,
                                                                 "sequence": "ACGTACGT",
                                                                 "keep": True })

### /neighbours/<sample_name>/<distance>

Get a array of tuples [[neighbour_name, distance]] of neighbours of sample_name up to the SNP cut-off distance.

    >>> requests.get("http://localhost:5000/neighbours/sample_name/20").json()

### /add_samples_from_mfsl

Add samples to catwalk in the multifasta singleline format.

    >>> requests.post("http://localhost:5000/add_samples_from_mfsl", json={"filepath": "mysamples.fa"})

The filepath must be readable from the catwalk server.

This is faster than sending the sequences over HTTP with /add_sample. For maximum performance, use fast storage such as SSD drives.

### /remove_sample/<sample_name>

Remove a sample from catwalk

    >>> requests.get("http://localhost:5000/remove_sample/sample_name")

## Benchmarks

Please see `benchmark` directory readme.

## Publications
A publication on CatWalk is in progress [ADD MedRxiv link - TODO]
